modules/storages/app/services/storages/file_link_sync_service.rb

```diff
@@ -0,0 +1,218 @@
+  def initialize(user:)
+    @user = user
+    @service_result = ServiceResult.success(result: [])
+  end
+
+  # Synchronize the "cached" FileLinks in the database with a Nextcloud response.
+  # We assume a OAuthClientToken is present, because this service is called after
+  # the storage endpoint.
+  # @return ServiceResult with FileLinks in the result
+  def call(file_links)
+    @file_links = file_links
+    storage_file_link_hash = @file_links.group_by(&:storage_id)
+    storage_file_link_hash.each do |storage_id, storage_file_links|
+      # Add new types of storages here. We currently only support Nextcloud
+      sync_nextcloud(storage_id, storage_file_links)
+    end
+
+    @service_result
+  end
+
+  def set_file_link_permissions(file_links, parsed_response)
+    file_links.each do |file_link|
+      origin_file_info_hash = parsed_response[file_link.origin_id]
+
+      case origin_file_info_hash['statuscode']
+      when 200 # Successfully found - update infos
+        next if origin_file_info_hash['trashed'] # moved to trash in Nextcloud
+
+        sync_single_file(file_link, origin_file_info_hash)
+        file_link.origin_permission = :view
+      when 403 # Forbidden - file from other user
+        file_link.origin_permission = :not_allowed
+      when 404 # Not found - permanently deleted in Nextcloud
+        file_link.destroy
+        next # don't save and don't add file_link to result!
+      else
+        file_link.origin_permission = :error
+      end
+
+      @service_result.result << file_link
+      file_link.save # Only saves to database if some field has actually changed.
+    end
+  end
+
+  # Write the updated information from Nextcloud to a single file
+  # @param storage Storage of the file
+  # @param origin_file_id Nextcloud ID of the file
+  # @param origin_file_info_hash Hash with updated information from Nextcloud
+  # "24" => {
+  #    "id" : 24,                 # origin_file_id
+  #    "ctime" : 0,               # Linux epoch file creation +overwrite
+  #    "mtime" : 1655301278,      # Linux epoch file modification +overwrite
+  #    "mimetype" : "application/pdf",  # +overwrite
+  #    "name" : "Nextcloud Manual.pdf", # "Canonical" name, could changed by owner +overwrite
+  #    "owner_id" : "admin",      # ID at Nextcloud side +overwrite
+  #    "owner_name" : "admin",    # Name at Nextcloud side +overwrite
+  #    "size" : 12706214,         # Not used yet in OpenProject +overwrite
+  #    "status" : "OK",           # Not used yet
+  #    "statuscode" : 200,        # Not used yet
+  #    "trashed" : false          # Exclude trashed files from result array of FileLinks
+  # }
+  # In case of permission errors (depending on the current user) we get:
+  # "24" => { "status" => "Forbidden", "statuscode" => 403 }
+  # In case of completely deleted file or other errors we might also get:
+  # "24" = { "status" => "Not Found", "statuscode" => 404 }
+  def sync_single_file(file_link, origin_file_info_hash)
```

Comment:
I'd basically agree with you. However, Wieland and I explicitly chose to write a separate test for this method in order to reduce the test complexity of the rest of the service. 
The clean option would be to convert this method into a separate service. However, this new service would again be kind of "public" and a increase of complexity for 5 lines of code.

What do you think?

Reply:
Honestly, this function is so basic, I rather wouldn't test it separately at all. I would like to have it private, because there is no need to allow developers using this outside of this class.